include/kb_client.hpp

```diff
@@ -59,194 +59,228 @@ std::map<msgpack::type::object_type, std::string> msgpack_type_map = {
+    if (type_string == "int32_t" && std::is_same<T, int32_t>::value) return true;
+    if (type_string == "int16_t" && std::is_same<T, int16_t>::value) return true;
+    if (type_string == "int8_t" && std::is_same<T, int8_t>::value) return true;
+    if (type_string == "float" && std::is_same<T, float>::value) return true;
+    if (type_string == "double" && std::is_same<T, double>::value) return true;
+    return (type_string == "bool" && std::is_same<T, bool>::value);
 }
 
 /*
- * A container hold a msgpack::object for deferred unpack.
+ * Abstract class for MsgpackObject and Array.
  */
 class Object {
-    // msgpack::object has a shallow copy constructor
-    msgpack::object value_;
+
+protected:
+    // size of the flattened array, 0 for scalar data and NIL
     std::size_t size_;
+    // data type, if container, it refers to the data type in the container
     std::string dtype_;
 
 public:
-    Object() = default;  // must be default constructable
-
-    explicit Object(const msgpack::object& value):
-        value_(value),
-        size_(
-            [&value]() -> std::size_t {
-                if (value.type == msgpack::type::object_type::NIL) return 0;
-                if (value.type == msgpack::type::object_type::ARRAY ||
-                        value.type == msgpack::type::object_type::MAP ||
-                        value.type == msgpack::type::object_type::BIN)
-                    return value.via.array.size;
-                return 1;
-                }()
-            ),
-        dtype_(msgpack_type_map.at(value.type))
-    {}
+    Object() : size_(0) {};
+    virtual ~Object() = default;
+
+    virtual std::size_t size() const = 0;
+    virtual std::string dtype() const = 0;
+    // empty vector for scalar data
+    virtual std::vector<std::size_t> shape() const = 0;
+    // empty string for scalar data
+    virtual std::string containerType() const = 0;
+};
+
+/*
+ * A container that holds a msgpack::object for deferred unpack.
+ */
+class MsgpackObject : public Object {
+
+    msgpack::object value_; // msgpack::object has a shallow copy constructor
+
+public:
+    MsgpackObject() = default;  // must be default constructable
 
-    ~Object() = default;
+    explicit MsgpackObject(const msgpack::object& value): value_(value) {
+        if (value.type == msgpack::type::object_type::ARRAY ||
+                value.type == msgpack::type::object_type::MAP ||
+                value.type == msgpack::type::object_type::BIN)
+            size_ = value.via.array.size;
+
+        if (value.type == msgpack::type::object_type::ARRAY)
+            dtype_ = msgpack_type_map.at(value.via.array.ptr[0].type);
+        else if (value.type == msgpack::type::object_type::BIN)
+            dtype_ = "char";
+        else if (value.type == msgpack::type::object_type::MAP
+            || value.type == msgpack::type::object_type::EXT)
+            dtype_ = "undefined";
```

Comment:
we can maybe give better information here. 
EXT is a user defined type, maybe we can say that and return the type ID for this object? `custom type ID:1`

Reply:
I am not sure whether I fully understood what you mean. In the specification, it says "Ext format family stores a tuple of an integer and a byte array."

Since all the data come from the upstream bridge. We may never use EXT to pack data. In the worst scenario, if we did use EXT to pack data one day and the karabo-bridge-cpp were not updated accordingly, the code would not crash and the showNext() would remind the user to check the unexpected data type. Then the user will inform us.